### PR TITLE
reconnect: Rework to allow real reconnecting

### PR DESCRIPTION
**Pull request by LucioFranco**
Reopened from #386

Signed-off-by: Lucio Franco <luciofranco14@gmail.com>

### CODE REVIEW

File: tower/src/lib.rs [hunk 1]
```diff
@@ -51,8 +51,12 @@ pub mod builder;
 pub mod layer;
 
 #[cfg(feature = "util")]
+#[doc(inline)]
 pub use self::util::{service_fn, ServiceExt};
 
+#[doc(inline)]
```
Comment (jonhoo) [inline tower/src/lib.rs:57]: Why inline these?

File: tower/src/reconnect/future.rs [hunk 2]
```diff
@@ -9,25 +9,51 @@ use std::{
...
     F: Future<Output = Result<T, E>>,
     E: Into<Error>,
+    ME: Into<Error>,
 {
     type Output = Result<T, Error>;
 
+    #[project]
     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        self.project().inner.poll(cx).map_err(Into::into)
+        //self.project().inner.poll(cx).map_err(Into::into)
```
Comment (jonhoo) [inline tower/src/reconnect/future.rs:48]: Did you mean to remove this comment?

File: tower/src/reconnect/mod.rs [hunk 3]
```diff
@@ -1,8 +1,21 @@
-#![allow(missing_docs)]
+//! Reconnect services when they fail.
+//!
+//! Reconnect takes some [`MakeService`] and will transform it into a
```
Comment (jonhoo) [inline tower/src/reconnect/mod.rs:3]:
```suggestion
//! Reconnect takes some [`MakeService`] and transforms it into a
```

File: tower/src/reconnect/mod.rs [hunk 4]
```diff
@@ -1,8 +1,21 @@
...
+//! [`Service`]. It will then attempt to lazily connect and
```
Comment (jonhoo) [inline tower/src/reconnect/mod.rs:4]:
```suggestion
//! [`Service`]. It then attempts to lazily connect and
```

File: tower/src/lib.rs (hunk 1 above)
Comment (LucioFranco) [reply tower/src/lib.rs:57]: I find the docs look much better 🤷

File: tower/src/reconnect/future.rs (hunk 2 above)
Comment (LucioFranco) [reply tower/src/reconnect/future.rs:48]: oops I just copied from the old PR :D

**Pull request merged**